File: android/app/src/main/java/com/openburnbar/data/support/BugReportService.kt
package com.openburnbar.data.support

import android.util.Log
import com.google.firebase.functions.FirebaseFunctions
import com.openburnbar.data.models.BugReportSubmission
import com.openburnbar.data.models.BugReportSubmissionResult
import kotlinx.coroutines.tasks.await

class BugReportService(
    private val functionsProvider: () -> FirebaseFunctions = {
        FirebaseFunctions.getInstance("us-central1")
    },
) {
    companion object {
        private const val TAG = "BugReportService"
    }

    suspend fun submit(submission: BugReportSubmission): Result<BugReportSubmissionResult> =
        runCatching {
            val payload =
                mutableMapOf<String, Any>(
                    "title" to submission.title,
                    "description" to submission.description,
                    "platform" to submission.platform,
                    "autoDispenseCLI" to submission.autoDispenseCLI,
                )

            submission.appVersion?.let { payload["appVersion"] = it }
            submission.osVersion?.let { payload["osVersion"] = it }
            submission.deviceModel?.let { payload["deviceModel"] = it }
            submission.diagnostics?.let { payload["diagnostics"] = it }
            submission.logsSnippet?.let { payload["logsSnippet"] = it }
            submission.requestedRuntime?.let { payload["requestedRuntime"] = it }
            submission.targetProject?.let { payload["targetProject"] = it }

            Log.i(TAG, "Submitting Android bug report: '${submission.title}'")

            val functions = functionsProvider()
            val result = functions.getHttpsCallable("submitBugReport").call(payload).await()
            val data = result.getData() as? Map<*, *> ?: error("Invalid response from server.")

            val reportId = data["reportId"] as? String ?: ""
            val missionId = data["missionId"] as? String
            val linear = data["linearIssue"] as? Map<*, *> ?: emptyMap<String, Any>()
            val identifier = linear["identifier"] as? String ?: "BB-ISSUE"
            val url = linear["url"] as? String ?: "https://linear.app"
            val isMock = linear["mock"] as? Boolean ?: false

            Log.i(TAG, "Bug report submitted successfully. Linear issue: $identifier")

            BugReportSubmissionResult(
                reportId = reportId,
                linearIdentifier = identifier,
                linearUrl = url,
                isMock = isMock,
                missionId = missionId,
            )
        }
}

File: AgentLens/Services/Support/BugReportService.swift
import Foundation
#if canImport(AppKit)
import AppKit
#endif
@preconcurrency import FirebaseFunctions
import OSLog

public struct BugReportSubmission: @unchecked Sendable {
    public let title: String
    public let description: String
    public let platform: String
    public let appVersion: String?
    public let osVersion: String?
    public let deviceModel: String?
    public let diagnostics: [String: Any]?
    public let logsSnippet: String?
    public let screenshotBase64: String?
    public let autoDispenseCLI: Bool
    public let requestedRuntime: String?
    public let targetProject: String?

    public init(
        title: String,
        description: String,
        platform: String = "macOS",
        appVersion: String? = nil,
        osVersion: String? = nil,
        deviceModel: String? = nil,
        diagnostics: [String: Any]? = nil,
        logsSnippet: String? = nil,
        screenshotBase64: String? = nil,
        autoDispenseCLI: Bool = true,
        requestedRuntime: String? = nil,
        targetProject: String? = nil
    ) {
        self.title = title
        self.description = description
        self.platform = platform
        self.appVersion = appVersion
        self.osVersion = osVersion
        self.deviceModel = deviceModel
        self.diagnostics = diagnostics
        self.logsSnippet = logsSnippet
        self.screenshotBase64 = screenshotBase64
        self.autoDispenseCLI = autoDispenseCLI
        self.requestedRuntime = requestedRuntime
        self.targetProject = targetProject
    }
}

public struct BugReportSubmissionResult: Sendable, Equatable {
    public let reportId: String
    public let linearIdentifier: String
    public let linearUrl: String
    public let isMock: Bool
    public let missionId: String?

    public init(
        reportId: String,
        linearIdentifier: String,
        linearUrl: String,
        isMock: Bool = false,
        missionId: String? = nil
    ) {
        self.reportId = reportId
        self.linearIdentifier = linearIdentifier
        self.linearUrl = linearUrl
        self.isMock = isMock
        self.missionId = missionId
    }
}

public enum BugReportService {
    private static let logger = Logger(subsystem: "com.openburnbar.app", category: "BugReportService")

    public static func submit(
        _ submission: BugReportSubmission,
        functions: Functions = Functions.functions(region: "us-central1")
    ) async throws -> BugReportSubmissionResult {
        var payload: [String: Any] = [
            "title": submission.title,
            "description": submission.description,
            "platform": submission.platform,
            "autoDispenseCLI": submission.autoDispenseCLI,
        ]

        if let v = submission.appVersion { payload["appVersion"] = v }
        if let v = submission.osVersion { payload["osVersion"] = v }
        if let v = submission.deviceModel { payload["deviceModel"] = v }
        if let d = submission.diagnostics { payload["diagnostics"] = d }
        if let l = submission.logsSnippet { payload["logsSnippet"] = l }
        if let s = submission.screenshotBase64 { payload["screenshotBase64"] = s }
        if let r = submission.requestedRuntime { payload["requestedRuntime"] = r }
        if let t = submission.targetProject { payload["targetProject"] = t }

        logger.info("Submitting bug report: '\(submission.title, privacy: .public)' (platform: \(submission.platform, privacy: .public))")

        let result = try await functions.httpsCallable("submitBugReport").call(payload)
        guard let data = result.data as? [String: Any] else {
            throw NSError(domain: "BugReportService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid response from server."])
        }

        let reportId = data["reportId"] as? String ?? ""
        let missionId = data["missionId"] as? String
        let linear = data["linearIssue"] as? [String: Any] ?? [:]
        let identifier = linear["identifier"] as? String ?? "BB-ISSUE"
        let url = linear["url"] as? String ?? "https://linear.app"
        let isMock = linear["mock"] as? Bool ?? false

        logger.info("Bug report submitted successfully. Linear issue: \(identifier, privacy: .public) (reportId: \(reportId, privacy: .public))")

        return BugReportSubmissionResult(
            reportId: reportId,
            linearIdentifier: identifier,
            linearUrl: url,
            isMock: isMock,
            missionId: missionId
        )
    }
}
